### ISSUE TITLE
[MODULE TESTING]  Addition of custom modules to the upstream `module_db`

### ISSUE THREAD

**Issue by anubhavjana** [torch-spyre/torch-spyre#1439]
We should be able to specify module in the config and if that module is not listed in the upstream `module_db`, we should be able to add / register that module in module_db and use it in the test framework. 

We should be able to do a patching before the `super.instantiate_test()` gets called in `spyre_test_base_common.py`

Please take a look at `spyre_upstream_patcher.py` for reference implementations for upstream ops_db patching with ops_list and dtype.

**Comment by AnishPahilajani** [comment]
As I understand it this will mean writing tests per module, and this is also proposing a completely separate design.
So I think your proposed workflow is the following

step 1
in `tests/test_suite_config.yaml`
```
    # Custom module registration for Granite
    supported_modules:
      - name: "GraniteAttention"
        module_path: "transformers.models.granite.modeling_granite.GraniteAttention"
        input_generator: "model_tests.module_input_generators.generate_granite_attention_inputs"
        dtypes:
          - float16
          - float32

```

Step 2.
Register custom modules in the upstream module_db. This would be done by adding ModuleInfo instances for each custom module. The modules are only added if they don't already exist in the database. 

so in `tests/spyre_test_base_common.py`
```
if cls.CUSTOM_MODULES:
            _SpyreModulePatcher(cls.CUSTOM_MODULES).patch()
```

similar to `spyre_upstream_patcher.py` I create `spyre_module_patcher.py `

```
class _SpyreModulePatcher:
    def __init__(self, modules_to_register):
        # Store what to patch
        
    def patch(self):
        # Find upstream object (module_db)
        # Check if module exists
        # Create ModuleInfo
        # Append to module_db in-place

```


I would also have to create an input generator

for example in a file called module_input_generators.py
```
def generate_granite_attention_inputs(
    module_cls, device, dtype, requires_grad=False, **kwargs
) -> Tuple[Any, Tuple, dict]:
    """Generate inputs for GraniteAttention."""
    from transformers.models.granite.configuration_granite import GraniteConfig
    
    config = GraniteConfig(
        hidden_size=768,
        num_attention_heads=12,
        num_key_value_heads=12,
        _attn_implementation="eager",  # Required to avoid KeyError
    )
    
    module = module_cls(config, layer_idx=0).to(device=device, dtype=dtype)
    module.eval()
    
    batch_size, seq_len = 2, 128
    hidden_states = torch.randn(
        batch_size, seq_len, config.hidden_size,
        device=device, dtype=dtype, requires_grad=requires_grad
    )
    
    # Create position embeddings (cos, sin) for rotary embeddings
    # Shape should be (batch_size, seq_len, head_dim)
    head_dim = config.hidden_size // config.num_attention_heads
    
    # Generate cos and sin embeddings with correct shape
    # The rotary embedding expects (batch, seq_len, head_dim)
    cos = torch.randn(batch_size, seq_len, head_dim, device=device, dtype=dtype)
    sin = torch.randn(batch_size, seq_len, head_dim, device=device, dtype=dtype)
    position_embeddings = (cos, sin)
    
    # Create attention mask (all ones for no masking)
    attention_mask = torch.ones(batch_size, 1, seq_len, seq_len, device=device, dtype=dtype)
    
    return module, (hidden_states,), {
        "position_embeddings": position_embeddings,
        "attention_mask": attention_mask
    }

.
.
.
def generate_granite_decoder_layer_inputs()
.
def generate_granite_mlp_inputs()
```


And then create file for testing all granite layers

**Comment by anubhavjana** [comment]
https://github.com/torch-spyre/torch-spyre/issues/1439#issuecomment-4212772497 - Yes, something like this. We can discuss the flow in details in the upcoming call. 

Basically, we are trying to bring up a common framework for everything so that it can be controlled by a single config. 

cc: @ashokponkumar

**Comment by ani300** [comment]
After discussion with @AnishPahilajani @anubhavjana and @romitjain, here's an updated plan:

1. Model info generation:
   1. First we load the HF model with `AutoModel.from_pretrained()` and move it to spyre/fp16
   2. Then we iterate through all the modules in the model through a `for module in model.modules():`
   3. For each module we install a forward hook through `module.register_module_forward_hook` [(doc)](https://docs.pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html). The hook will capture the module class path so it can be instantiated later, together with each input's name and shape/stride/SpyreTensorLayout (when that's possible). That information will be saved to a global model_info dictionary.
   4. We run the model forward pass with an array of shapes we care about (prefill, decode, batch, etc.). This will fill up the model_info dictionary with all the information we care about.
   5. We write the model_info dictionary into a YAML file, which can then be combined through any utility with all the other models to have a single YAML file.

2. When generating ModuleInfo for the module_db, we should read the one yaml file with all the models info in there and append a new ModuleInfo for each module. To create the input generator for that module, we can programatiaclly define a new function composed of a list of pre-written input generators for each kind of input we care about (tokens, logits, hidden_states, mask, etc.).
3. We might want to add some specific to Spyre tests for each module we care about. Those should be installed as an extension to TestModule class in https://github.com/pytorch/pytorch/blob/main/test/test_modules.py. We should base those on testing specific SpyreTensorLayouts are present, or any other Spyre-specific stuff we might care about (compilation maybe?)
